egs/librispeech/asr/simple_v1/asg_train.py

```diff
@@ -0,0 +1,425 @@
+        where ok_frames is the frames for successful (finite) segments, and
+       all_frames is the frames for all segments (finite or not).
+    '''
+    mask = torch.ne(tot_scores, -math.inf)
+    # finite_indexes is a tensor containing successful segment indexes, e.g.
+    # [ 0 1 3 4 5 ]
+    finite_indexes = torch.nonzero(mask).squeeze(1)
+    if False:
+        bad_indexes = torch.nonzero(~mask).squeeze(1)
+        if bad_indexes.shape[0] > 0:
+            print("Bad indexes: ", bad_indexes, ", bad lengths: ",
+                  frames_per_seq[bad_indexes], " vs. max length ",
+                  torch.max(frames_per_seq), ", avg ",
+                  (torch.sum(frames_per_seq) / frames_per_seq.numel()))
+    # print("finite_indexes = ", finite_indexes, ", tot_scores = ", tot_scores)
+    ok_frames = frames_per_seq[finite_indexes].sum()
+    all_frames = frames_per_seq.sum()
+    return (tot_scores[finite_indexes].sum(), ok_frames, all_frames)
+
+
+def get_objf(batch: Dict,
+             model: AcousticModel,
+             P: k2.Fsa,
+             device: torch.device,
+             graph_compiler: AsgTrainingGraphCompiler,
+             is_training: bool,
+             optimizer: Optional[torch.optim.Optimizer] = None):
+    feature = batch['features']
+    supervisions = batch['supervisions']
+    supervision_segments = torch.stack(
+        (supervisions['sequence_idx'],
+         torch.floor_divide(supervisions['start_frame'],
+                            model.subsampling_factor),
+         torch.floor_divide(supervisions['num_frames'],
+                            model.subsampling_factor)), 1).to(torch.int32)
+    indices = torch.argsort(supervision_segments[:, 2], descending=True)
+    supervision_segments = supervision_segments[indices]
+
+    texts = supervisions['text']
+    texts = [texts[idx] for idx in indices]
+    assert feature.ndim == 3
+    # print(supervision_segments[:, 1] + supervision_segments[:, 2])
+
+    feature = feature.to(device)
+    # at entry, feature is [N, T, C]
+    feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+    if is_training:
+        nnet_output = model(feature)
+    else:
+        with torch.no_grad():
+            nnet_output = model(feature)
+
+    # nnet_output is [N, C, T]
+    nnet_output = nnet_output.permute(0, 2, 1)  # now nnet_output is [N, T, C]
+
+    if is_training:
+        num, den = graph_compiler.compile(texts, P)
+    else:
+        with torch.no_grad():
+            num, den = graph_compiler.compile(texts, P)
+
+    assert num.requires_grad == is_training
+    assert den.requires_grad is False
+    num = num.to(device)
+    den = den.to(device)
+
+
+    # nnet_output2 = nnet_output.clone()
+    # blank_bias = -7.0
+    # nnet_output2[:,:,0] += blank_bias
+
+    dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+    assert nnet_output.device == device
+    # TODO(haowen): with a small `beam`, we may get empty `target_graph`,
```

Comment:
it's better if we can delete those outdated comments as we have filtered `inf` even for decode scripts now (for pruned version), of course we don't need this for non-pruned version.

Reply:
fixed.